File: frontend/apps/ios/Pantopus/Features/Homes/HomeDashboardViewModel.swift
//
//  HomeDashboardViewModel.swift
//  Pantopus
//
//  Fetches, concurrently:
//   - `GET /api/homes/:id` (detail, with a `/public-profile` fallback),
//   - `GET /api/homes/:id/dashboard`   — hero stats + Overview sections,
//   - `GET /api/homes/:id/health-score`,
//   - `GET /api/homes/:id/seasonal-checklist`,
//   - `GET /api/homes/:id/property-value`,
//   - `GET /api/homes/:id/bill-trends`.
//
//  Each Home Intelligence card owns its own state so one failing read
//  can't blank the screen.
//
import Foundation
import Observation

/// Projection of the home header + stats + tab strip.
public struct HomeDashboardContent: Sendable {
    public let address: String
    /// True when the home has any verified owner; drives the header
    /// "Verified" badge and the summary status row. Distinct from
    /// `isVerifiedOwner` because the home can have a verified owner
    /// who isn't the signed-in user.
    public let verified: Bool
    /// True when the signed-in user is the verified owner of this home.
    /// Drives the claim-ownership banner gate: shown when this is false
    /// regardless of whether anyone else is a verified owner.
    public let isVerifiedOwner: Bool
    public let stats: [HomeHeroStat]
    public let quickActions: [QuickActionTile]
    public let tabs: [GridTabsTab]
    public let overview: HomeDashboardOverviewContent
    public let attentionSummary: HomeDashboardAttentionSummary?
}

public struct HomeDashboardOverviewContent: Sendable {
    public let upcoming: [HomeDashboardTimelineItem]
    public let activity: [HomeDashboardActivityItem]
    public let emergency: HomeDashboardEmergencyInfo
}

public struct HomeDashboardTimelineItem: Sendable, Identifiable {
    public let id: String
    public let icon: PantopusIcon
    public let tone: QuickActionTone
    public let title: String
    public let subtitle: String
    public let trailing: String?
}

public struct HomeDashboardActivityItem: Sendable, Identifiable {
    public let id: String
    public let initials: String
    public let tone: QuickActionTone
    public let title: String
    public let detail: String
    public let time: String
}

public struct HomeDashboardEmergencyInfo: Sendable {
    public let title: String
    public let body: String
    public let isConfigured: Bool
}

public struct HomeDashboardAttentionSummary: Sendable {
    public let message: String
    public let chips: [HomeDashboardQuickJump]
}

public struct HomeDashboardQuickJump: Sendable, Identifiable {
    public let id: String
    public let label: String
    public let icon: PantopusIcon
    public let actionId: String
}

public struct HomeDashboardBrandNewContent: Sendable {
    public let content: HomeDashboardContent
    public let onboardingSteps: [HomeDashboardOnboardingStep]
}

public struct HomeDashboardOnboardingStep: Sendable, Identifiable {
    public let id: String
    public let title: String
    public let body: String
    public let cta: String
    public let icon: PantopusIcon
    public let tone: QuickActionTone
    public let actionId: String
}

/// Observed state for the Home Dashboard screen.
public enum HomeDashboardState: Sendable {
    case loading
    case loaded(HomeDashboardContent)
    case empty(HomeDashboardBrandNewContent)
    case needsAttention(HomeDashboardContent)
    case error(message: String)
}

/// Per-card state for the Home Intelligence stack. Each card renders its
/// own loading / loaded / absent / error surface so a failure in one read
/// never blanks the dashboard.
public enum HomeIntelligenceCardState<Value: Sendable>: Sendable {
    case loading
    case loaded(Value)
    /// The signed-in member isn't permitted to see this card (HTTP 403).
    case forbidden
    case failed(message: String)

    public var value: Value? {
        if case let .loaded(value) = self { return value }
        return nil
    }

    public var isLoading: Bool {
        if case .loading = self { return true }
        return false
    }
}

/// Backs [`HomeDashboardView`].
@Observable
@MainActor
final class HomeDashboardViewModel {
    /// Currently displayed state.
    private(set) var state: HomeDashboardState = .loading
    /// Currently selected grid tab.
    var selectedTab: String = "overview"

    // MARK: - Home Intelligence (independent per-card state)

    private(set) var healthScore: HomeIntelligenceCardState<HomeHealthScoreDTO> = .loading
    private(set) var checklist: HomeIntelligenceCardState<SeasonalChecklistDTO> = .loading
    private(set) var propertyValue: HomeIntelligenceCardState<HomePropertyValueDTO> = .loading
    private(set) var billTrends: HomeIntelligenceCardState<HomeBillTrendsDTO> = .loading
    /// Checklist item ids with an in-flight PATCH — the row disables while
    /// its mutation is awaiting the server's returned item state.
    private(set) var pendingChecklistItemIds: Set<String> = []

    private let homeId: String
    private let api: APIClient

    // Raw responses; `rebuild()` composes the rendered content from them.
    private var detailData: HomeDetail?
    private var publicData: HomePublicProfileResponse.HomePublicProfile?
    private var dashboardData: HomeDashboardResponse?

    init(homeId: String, api: APIClient = .shared) {
        self.homeId = homeId
        self.api = api
    }

    /// Initial load; no-op when we already have content.
    func load() async {
        if isContentState { return }
        if let sampleState = HomeDashboardSampleData.state(for: homeId) {
            state = sampleState
            return
        }
        state = .loading
        await fetchAll()
    }

    /// Pull-to-refresh / retry.
    func refresh() async {
        if let sampleState = HomeDashboardSampleData.state(for: homeId) {
            state = sampleState
            return
        }
        await fetchAll()
    }

    private var isContentState: Bool {
        switch state {
        case .loaded, .empty, .needsAttention:
            true
        case .loading, .error:
            false
        }
    }

    // MARK: - Fetch

    private func fetchAll() async {
        async let core: Void = fetchCore()
        async let health: Void = loadHealthScore()
        async let seasonal: Void = loadChecklist()
        async let property: Void = loadPropertyValue()
        async let trends: Void = loadBillTrends()
        await core
        await health
        await seasonal
        await property
        await trends
    }

    /// Home detail (identity / ownership) + the dashboard aggregate.
    private func fetchCore() async {
        async let detailOutcome = loadDetail()
        async let dashboard = loadDashboard()
        let outcome = await detailOutcome
        dashboardData = await dashboard

        switch outcome {
        case let .detail(home):
            detailData = home
            publicData = nil
            rebuild()
        case .needsPublicProfile:
            await fetchPublicProfile()
        case let .failed(message):
            detailData = nil
            publicData = nil
            state = .error(message: message)
        }
    }

    private enum DetailOutcome {
        case detail(HomeDetail)
        case needsPublicProfile
        case failed(message: String)
    }

    private func loadDetail() async -> DetailOutcome {
        do {
            let response: HomeDetailResponse = try await api.request(HomesEndpoints.detail(homeId: homeId))
            return .detail(response.home)
        } catch APIError.forbidden, APIError.notFound {
            return .needsPublicProfile
        } catch {
            return .failed(message: (error as? APIError)?.errorDescription ?? "Couldn't load home.")
        }
    }

    private func loadDashboard() async -> HomeDashboardResponse? {
        try? await api.request(
            HomeDashboardEndpoints.dashboard(homeId: homeId),
            as: HomeDashboardResponse.self
        )
    }

    private func fetchPublicProfile() async {
        do {
            let response: HomePublicProfileResponse =
                try await api.request(HomesEndpoints.publicProfile(homeId: homeId))
            publicData = response.home
            detailData = nil
            rebuild()
        } catch {
            state = .error(message: (error as? APIError)?.errorDescription ?? "Couldn't load home.")
        }
    }

    // MARK: - Home Intelligence reads

    /// Mirrors RN's `useHomeIntelligence`, which always forces a server
    /// recompute so a stale zero-score can't mask a populated home.
    private func loadHealthScore() async {
        healthScore = await fetchCard {
            try await self.api.request(
                HomeDashboardEndpoints.healthScore(homeId: self.homeId, force: true),
                as: HomeHealthScoreDTO.self
            )
        }
        // The Overview's emergency row reads the health breakdown.
        rebuild()
    }

    private func loadChecklist() async {
        checklist = await fetchCard {
            try await self.api.request(
                HomeDashboardEndpoints.seasonalChecklist(homeId: self.homeId),
                as: SeasonalChecklistDTO.self
            )
        }
    }

    private func loadPropertyValue() async {
        propertyValue = await fetchCard {
            try await self.api.request(
                HomeDashboardEndpoints.propertyValue(homeId: self.homeId),
                as: HomePropertyValueDTO.self
            )
        }
    }

    private func loadBillTrends() async {
        billTrends = await fetchCard {
            try await self.api.request(
                HomeDashboardEndpoints.billTrends(homeId: self.homeId),
                as: HomeBillTrendsDTO.self
            )
        }
    }

    private func fetchCard<Value: Sendable>(
        _ work: () async throws -> Value
    ) async -> HomeIntelligenceCardState<Value> {
        do {
            let value = try await work()
            return .loaded(value)
        } catch APIError.forbidden {
            return .forbidden
        } catch {
            return .failed(
                message: (error as? APIError)?.errorDescription ?? "Couldn't load this card."
            )
        }
    }

    // MARK: - Seasonal checklist actions

    /// `PATCH …/seasonal-checklist/:itemId { status: "completed" }`.
    func completeChecklistItem(_ itemId: String) async {
        await updateChecklistItem(itemId, status: "completed")
    }

    /// `PATCH …/seasonal-checklist/:itemId { status: "skipped" }`.
    func skipChecklistItem(_ itemId: String) async {
        await updateChecklistItem(itemId, status: "skipped")
    }

    /// The GET is idempotent-generate: it creates the current season's
    /// items when the home has none, so "Generate checklist" is a re-read.
    func generateChecklist() async {
        checklist = .loading
        await loadChecklist()
    }

    /// Re-reads only the health score (used after a checklist mutation and
    /// on card-level Retry).
    func refreshHealthScore() async {
        healthScore = .loading
        await loadHealthScore()
    }

    func retryPropertyValue() async {
        propertyValue = .loading
        await loadPropertyValue()
    }

    func retryBillTrends() async {
        billTrends = .loading
        await loadBillTrends()
    }

    private func updateChecklistItem(_ itemId: String, status: String) async {
        guard !pendingChecklistItemIds.contains(itemId) else { return }
        pendingChecklistItemIds.insert(itemId)
        defer { pendingChecklistItemIds.remove(itemId) }

        do {
            let updated: SeasonalChecklistItemDTO = try await api.request(
                HomeDashboardEndpoints.updateSeasonalChecklistItem(
                    homeId: homeId,
                    itemId: itemId,
                    status: status
                )
            )
            // Reflect exactly what the server returned, then re-read the
            // score (seasonal progress is one of its six dimensions).
            applyChecklistItem(updated)
            await loadHealthScore()
        } catch {
            checklist = .failed(
                message: (error as? APIError)?.errorDescription
                    ?? "Couldn't update that task. Try again."
            )
        }
    }

    /// Splice the server's returned row back into the loaded checklist and
    /// recompute progress the same way the backend does (`home.js:7526`).
    private func applyChecklistItem(_ updated: SeasonalChecklistItemDTO) {
        guard let current = checklist.value else { return }
        let items = current.items.map { $0.id == updated.id ? updated : $0 }
        let carryover = current.carryover.map { block in
            SeasonalChecklistCarryoverDTO(
                season: block.season,
                items: block.items.map { $0.id == updated.id ? updated : $0 }
            )
        }
        let completed = items.filter(\.isResolved).count
        checklist = .loaded(
            SeasonalChecklistDTO(
                season: current.season,
                items: items,
                progress: SeasonalChecklistProgressDTO(
                    total: items.count,
                    completed: completed,
                    percentage: items.isEmpty ? 0 : Int((Double(completed) / Double(items.count) * 100).rounded())
                ),
                carryover: carryover
            )
        )
    }

    // MARK: - Projection

    private func rebuild() {
        if let detailData {
            state = .loaded(content(
                address: detailData.base.address ?? detailData.base.name ?? "Home",
                // Header badge / summary row: home has any verified owner.
                verified: detailData.isOwner || detailData.owners.contains { $0.ownerStatus == "verified" },
                // Banner gate: I'm the verified owner only when isOwner is
                // true and there's no pending claim still in flight.
                isVerifiedOwner: detailData.isOwner && !detailData.isPendingOwner
            ))
        } else if let publicData {
            state = .loaded(content(
                address: publicData.address,
                verified: publicData.hasVerifiedOwner,
                // Public-profile path is hit when the user is NOT a verified
                // owner; the private detail call returned 403/404 first.
                isVerifiedOwner: false
            ))
        }
    }

    private func content(
        address: String,
        verified: Bool,
        isVerifiedOwner: Bool
    ) -> HomeDashboardContent {
        let counts = dashboardData?.counts
        return HomeDashboardContent(
            address: address,
            verified: verified,
            isVerifiedOwner: isVerifiedOwner,
            stats: HomeDashboardProjection.stats(counts: counts),
            quickActions: HomeDashboardProjection.quickActions(counts: counts),
            tabs: HomeDashboardProjection.tabs,
            overview: HomeDashboardProjection.overview(
                dashboard: dashboardData,
                health: healthScore.value
            ),
            attentionSummary: nil
        )
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/homes/HomeDashboardViewModel.kt
@file:Suppress("MagicNumber")

package app.pantopus.android.ui.screens.homes

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.api.models.homes.HomeDetail
import app.pantopus.android.data.api.models.homes.HomePublicProfile
import app.pantopus.android.data.api.net.NetworkError
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.homes.HomesRepository
import app.pantopus.android.ui.screens.shared.content_detail.GridTabsTab
import app.pantopus.android.ui.screens.shared.content_detail.HomeHeroStat
import app.pantopus.android.ui.screens.shared.content_detail.QuickActionTile
import app.pantopus.android.ui.screens.shared.content_detail.QuickActionTone
import app.pantopus.android.ui.theme.PantopusIcon
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

/** Key used to read the home id from the nav backstack's SavedStateHandle. */
const val HOME_DASHBOARD_HOME_ID_KEY = "homeId"

/** Projection shown by [HomeDashboardScreen]. */
data class HomeDashboardContent(
    val address: String,
    /**
     * True when the home has any verified owner; drives the header
     * "Verified" badge and the summary status row. Distinct from
     * [isVerifiedOwner] because the home can have a verified owner who
     * isn't the signed-in user.
     */
    val verified: Boolean,
    /**
     * True when the signed-in user is the verified owner of this home.
     * Drives the claim-ownership banner gate: shown when this is false
     * regardless of whether anyone else is a verified owner.
     */
    val isVerifiedOwner: Boolean,
    val stats: List<HomeHeroStat>,
    val quickActions: List<QuickActionTile>,
    val tabs: List<GridTabsTab>,
    val overview: HomeDashboardOverviewContent,
    val attentionSummary: HomeDashboardAttentionSummary? = null,
)

data class HomeDashboardOverviewContent(
    val upcoming: List<HomeDashboardTimelineItem>,
    val activity: List<HomeDashboardActivityItem>,
    val emergency: HomeDashboardEmergencyInfo,
)

data class HomeDashboardTimelineItem(
    val id: String,
    val icon: PantopusIcon,
    val tone: QuickActionTone,
    val title: String,
    val subtitle: String,
    val trailing: String?,
)

data class HomeDashboardActivityItem(
    val id: String,
    val initials: String,
    val tone: QuickActionTone,
    val title: String,
    val detail: String,
    val time: String,
)

data class HomeDashboardEmergencyInfo(
    val title: String,
    val body: String,
    val isConfigured: Boolean,
)

data class HomeDashboardAttentionSummary(
    val message: String,
    val chips: List<HomeDashboardQuickJump>,
)

data class HomeDashboardQuickJump(
    val id: String,
    val label: String,
    val icon: PantopusIcon,
    val actionId: String,
)

data class HomeDashboardBrandNewContent(
    val content: HomeDashboardContent,
    val onboardingSteps: List<HomeDashboardOnboardingStep>,
)

data class HomeDashboardOnboardingStep(
    val id: String,
    val title: String,
    val body: String,
    val cta: String,
    val icon: PantopusIcon,
    val tone: QuickActionTone,
    val actionId: String,
)

/** Observed state for the Home Dashboard. */
sealed interface HomeDashboardUiState {
    data object Loading : HomeDashboardUiState

    data class Loaded(
        val content: HomeDashboardContent,
    ) : HomeDashboardUiState

    data class Empty(
        val brandNew: HomeDashboardBrandNewContent,
    ) : HomeDashboardUiState

    data class NeedsAttention(
        val content: HomeDashboardContent,
    ) : HomeDashboardUiState

    data class Error(
        val message: String,
    ) : HomeDashboardUiState
}

/**
 * ViewModel for the Home Dashboard screen. Receives the home id via the
 * nav-backstack [SavedStateHandle].
 */
@HiltViewModel
class HomeDashboardViewModel
    @Inject
    constructor(
        private val repo: HomesRepository,
        savedStateHandle: SavedStateHandle,
    ) : ViewModel() {
        private val homeId: String =
            requireNotNull(savedStateHandle[HOME_DASHBOARD_HOME_ID_KEY]) {
                "HomeDashboardViewModel requires a '$HOME_DASHBOARD_HOME_ID_KEY' nav arg."
            }

        private val _state = MutableStateFlow<HomeDashboardUiState>(HomeDashboardUiState.Loading)

        /** Observed state. */
        val state: StateFlow<HomeDashboardUiState> = _state.asStateFlow()

        private val _selectedTab = MutableStateFlow("overview")

        /** Currently-selected grid tab. */
        val selectedTab: StateFlow<String> = _selectedTab.asStateFlow()

        /** Switch the active grid tab. */
        fun selectTab(id: String) {
            _selectedTab.value = id
        }

        /** Expose the home id so the screen can build outbound nav routes. */
        fun currentHomeId(): String? = homeId

        /**
         * Display name of the loaded home, used as the 2-line top-bar
         * subtitle on the Access codes destination. Returns null while
         * the dashboard is still loading.
         */
        fun currentHomeName(): String? =
            when (val current = _state.value) {
                is HomeDashboardUiState.Loaded -> current.content.address
                is HomeDashboardUiState.Empty -> current.brandNew.content.address
                is HomeDashboardUiState.NeedsAttention -> current.content.address
                HomeDashboardUiState.Loading, is HomeDashboardUiState.Error -> null
            }

        /** Initial load; no-op when already loaded. */
        fun load() {
            if (_state.value is HomeDashboardUiState.Loaded ||
                _state.value is HomeDashboardUiState.Empty ||
                _state.value is HomeDashboardUiState.NeedsAttention
            ) {
                return
            }
            refresh()
        }

        /** Retry / pull-to-refresh. */
        fun refresh() {
            HomeDashboardSampleData.stateFor(homeId)?.let { sample ->
                _state.value = sample
                return
            }
            _state.value = HomeDashboardUiState.Loading
            viewModelScope.launch { fetch() }
        }

        private suspend fun fetch() {
            when (val result = repo.detail(homeId)) {
                is NetworkResult.Success -> applyDetail(result.data.home)
                is NetworkResult.Failure ->
                    if (result.error is NetworkError.Forbidden || result.error is NetworkError.NotFound) {
                        fetchPublic()
                    } else {
                        _state.value = HomeDashboardUiState.Error(result.error.message)
                    }
            }
        }

        private suspend fun fetchPublic() {
            when (val result = repo.publicProfile(homeId)) {
                is NetworkResult.Success -> applyPublic(result.data.home)
                is NetworkResult.Failure -> _state.value = HomeDashboardUiState.Error(result.error.message)
            }
        }

        private fun applyDetail(detail: HomeDetail) {
            val address = detail.address ?: detail.name ?: "Home"
            val stats =
                listOf(
                    HomeHeroStat("packages", "4", "Packages"),
                    HomeHeroStat("access_codes", "2", "Access codes"),
                    HomeHeroStat("tasks", "7", "Tasks"),
                )
            // Header / summary: home has any verified owner. Banner gate:
            // I'm the verified owner only when isOwner is true and there
            // is no pending claim still in flight.
            val homeVerified = detail.isOwner || detail.owners.any { it.ownerStatus == "verified" }
            val iAmVerified = detail.isOwner && !detail.isPendingOwner
            _state.value =
                HomeDashboardUiState.Loaded(
                    content(address, verified = homeVerified, isVerifiedOwner = iAmVerified, stats = stats),
                )
        }

        private fun applyPublic(publicProfile: HomePublicProfile) {
            val stats =
                listOf(
                    HomeHeroStat("packages", "0", "Packages"),
                    HomeHeroStat("access_codes", "0", "Access codes"),
                    HomeHeroStat("tasks", "0", "Tasks"),
                )
            // Public-profile path is hit when the user is NOT a verified
            // owner; the private detail call returned 403/404 first.
            _state.value =
                HomeDashboardUiState.Loaded(
                    content(
                        address = publicProfile.address,
                        verified = publicProfile.hasVerifiedOwner,
                        isVerifiedOwner = false,
                        stats = stats,
                    ),
                )
        }

        private fun content(
            address: String,
            verified: Boolean,
            isVerifiedOwner: Boolean,
            stats: List<HomeHeroStat>,
        ): HomeDashboardContent =
            HomeDashboardContent(
                address = address,
                verified = verified,
                isVerifiedOwner = isVerifiedOwner,
                stats = stats,
                quickActions =
                    HomeDashboardSampleData.populatedQuickActions,
                tabs = HomeDashboardSampleData.tabs,
                overview = HomeDashboardSampleData.populatedOverview,
                attentionSummary = null,
            )
    }
